core/src/main/scala/cats/Foldable.scala

```diff
@@ -433,4 +448,26 @@ object Foldable {
     }
     M.tailRecM(z)(go)
   }
+
+
+  private sealed abstract class Source[+A] {
+    def uncons: Option[(A, Source[A])]
+  }
+
+  private object Source {
+    val Empty: Source[Nothing] = new Source[Nothing] {
+      def uncons = None
+    }
+
+    val NowEmpty: Eval[Source[Nothing]] = Now(Empty)
+
+    def cons[A](a: A, src: Eval[Source[A]]): Source[A] = new Source[A] {
+      def uncons = Some((a, src.value))
```

Comment:
I'm a little nervous about these calls to value. Can we add a comment about the stack safety here? I think the answer is that we can't recursively call .value in an Eval and we should never be in an Eval when we call uncons, is that right?